finetuner/tuner/keras/__init__.py

```diff
@@ -173,24 +145,31 @@ def fit(
                 collate_fn=collate_fn,
             )
 
-        optimizer = optimizer or self._get_default_optimizer(learning_rate)
+        self.state = TunerState(num_epochs=epochs)
+        self._trigger_callbacks('on_fit_begin')
 
-        m_train_loss = ScalarSequence('train')
-        m_eval_loss = ScalarSequence('eval')
         with get_device(device):
             for epoch in range(epochs):
-                lt = self._train(
-                    train_dl,
-                    optimizer,
-                    description=f'Epoch {epoch + 1}/{epochs}',
-                )
-                m_train_loss += lt
+
+                # Setting here as re-shuffling can change number of batches
+                self.state.epoch = epoch
+                self.state.num_batches_train = train_dl.get_size()
+
+                self._trigger_callbacks('on_epoch_begin')
+
+                self._trigger_callbacks('on_train_begin')
```

Comment:
~i think `on_train_begin` should be moved out of the for loop, right? otherwise it's identical to `on_epoch_begin`~

now I understand, we have another `on_fit_begin`, then can we remove `on_train_begin` to something like `on_train_epoch_begin`?

Reply:
You mean to rename it? Because `on_train_begin` executes exactly at the start of each training epoch

Reply:
yes, how do you think, i find the keras naming very clear: 
![6619160E-3BD4-4F56-8DAD-C150696FED6E](https://user-images.githubusercontent.com/9794489/143462573-79d6e6e1-879e-4f2e-8a36-da0d5e833c88.png)

Reply:
Good point, I will rename it